File: python/InsertionSort.py
def insertion_sort_ascended(a):
    k =0
    for i in range(1, len(a)):
        if(a[i] < a[i-1]):  # 第i位之前已排好序，如果当前i位比i-1大，则已成升序，不用再排
            k = k+1
            for j in range(i, 0, -1):
                if(a[j] < a[j-1]):
                    a[j-1], a[j] = a[j], a[j-1]
            print("after %d change: "%k, a)
    return a

def insertion_sort_dascended(a):
    k =0
    for i in range(1, len(a)):
        if(a[i] > a[i-1]):  # 第i位之前已排好序，如果当前i位比i-1大，则已成降序，不用再排
            k = k+1
            for j in range(i, 0, -1):
                if(a[j] > a[j-1]):
                    a[j-1], a[j] = a[j], a[j-1]
            print("after %d change: "%k, a)
    return a

def main():
    Array = list(map(int,input("please input the array(split by space):").split()))
    print("input is: ", Array)
    SortedArrayAscend = insertion_sort_ascended(Array)
    print("ascend result is: ", SortedArrayAscend)

    SortedArrayDascend = insertion_sort_dascended(Array)
    print("dascend result is: ", SortedArrayDascend)


if __name__ == "__main__":
    main()

File: C++/InsertionSort.cpp
#include <iostream>
#include <stdlib.h>
#include <vector>

using namespace std;

void print_array(auto array)
{
    for(int i : array)
    {
        cout << i << " ";
    }
    cout << endl;
}

void insertion_sort(auto &array)
{
    auto len = array.size();
    int k = 0;
    for (int i = 1; i < len; i++)
    {
        if (array[i] < array[i-1])
        {
            for (int j = i; j > 0; j--)
            {
                if (array[j] < array[j-1])
                {
                    auto temp = array[j-1];
                    array[j-1] = array[j];
                    array[j] = temp;
                }
            }
            cout << "after " << ++k << " change: ";
            print_array(array);
        }
    }
}

int main()
{  
    vector<int> array;
    char ch;
    cout << "please input the array(split by space):";
    while((ch = cin.get()) != '\n')
    {
        if(ch != ' ')
            array.push_back(int(ch - '0'));
    }
    cout << "origin is: " ;
    print_array(array);

    insertion_sort(array);
    cout << "result is: " ;
    print_array(array);

    system("pause");
    return 0;
}